README.md
# Dynosaur

## coming soon...


docs/CHANGELOG
v0.10 [2015-11-02]
    - added events for twodns controls
    - changed button size (better for german language)

v0.09 [2015-11-02]
    - added 'lua51.dll'
    - added 'docs/LICENSE-Lua'
    - added vars to english lang file
    - changed: "system/modules/lualibs/socket/http.lua"
        - changed 'USERAGENT' to "Dynosaur"
    - changed ID names
    - removed unneeded IDs
    - moved control events to the bottom of his functions
    - code cleanup
    - using table namespaces for some functions

v0.08 [2015-10-31]
    - added multi-language support
        - added 'show_lang_dialog()' function
        - added 'show_lang_dialog_on_first_start()' function
        - added 'Change Language' menubar item with icon

v0.07 [2015-10-31]
    - twodns settings window:
        - added busy cursor feature while open the window
        - added statusbar
        - added elements to add an existing account
    - added 'trimstring()' function to 'util.lua', usage: util.trimstring( str )

v0.06 [2015-10-31]
    - added 'make_dbs()' funktion to create empty databases on start (if not exists)
    - changed 'make_logs()' function, check if log folder exists, if not then create it
    - changed 'make_dbs()' function, check if log folder exists, if not then create it

v0.05 [2015-10-30]
    - changed frame size
    - added aeslua/bitlib to encrypt account passwords before saving it to the table
    - added 3rd party licenses to '/docs'
    - added twodns item with icon to menubar
    - added noip item with icon to menubar
    - added dyndns item with icon to menubar

v0.04 [2015-10-29]
    - added minimize to tray functionality
    - show file size of logfile in log window
    - some log window improvements
    - using wxFile.Exists instead of LuaFileSystem lib
    - add icons to menubar items
    - added menubar item for: twodns log, noip log, dyndns log

v0.03 [2015-10-28]
    - added log window
    - write log handler functions: log.read(), log.write(), log.clear()
    - added 'make_logs()' function to create logfiles on start (if not exists)
    - added settings window
    - changed menubar

v0.02 [2015-10-27]
    - added menubar
    - added about window

v0.01 [2015-10-26]
    - writing basic app frontend

system/core/util.lua
--[[

    util.lua written by blastbeat and pulsar

    based on "luadch/core/util.lua"


    Description:

    table = util.loadtable( path )  -- loads a local table from file

    util.savearray( array, path )  -- saves an array to a local file

    util.savetable( tbl, name, path )  -- saves a table to a local file

    number, number, number, number = util.formatseconds( t )  -- converts time to: days, hours, minutes, seconds (for example: d, h, m, s = util.formatseconds( os.difftime( os.time( ), signal.get( "start" ) ) ) )

    string/nil = util.formatbytes( bytes )  -- returns converted bytes as a sting e.g. "209.81 GB"

    number = util.generatepass( len )  -- returns a random generated alphanumerical password with length = len; if no param is specified then len = 20

    number = util.date( )  -- returns current date in new luadch date style: yyyymmddhhmmss

    number, number, number, number, number/nil, err = util.difftime( t1, t2 )  -- returns difftime between two luadch date style values (new luadch date style) (for example : y, d, h, m, s = util.difftime( util.date(), 20140617031630 ) )

    number/nil, err = util.convertepochdate( t )  -- convert os.time() "epoch" date to luadch date style: yyyymmddhhmmss (as number)

]]


----------------------------
--[DEFINITION/DECLARATION]--
----------------------------

--// functions
local sortserialize
local savearray
local savetable
local loadtable
local formatseconds
local formatbytes
local generatepass
local date
local difftime
local convertepochdate
local trimstring

--// table lookups
local os_time = os.time
local os_date = os.date
local os_difftime = os.difftime
local math_floor = math.floor
local math_random = math.random
local math_randomseed = math.randomseed
local string_format = string.format


----------
--[CODE]--
----------

sortserialize = function( tbl, name, file, tab, r )
    tab = tab or ""
    local temp = { }
    for key, k in pairs( tbl ) do
        --if type( key ) == "string" or "number" then
            table.insert( temp, key )
        --end
    end
    table.sort( temp )
    local str = tab .. name
    if r then
        file:write( str .. " {\n\n" )
    else
        file:write( str .. " = {\n\n" )
    end
    for k, key in ipairs( temp ) do
        if ( type( tbl[ key ] ) ~= "function" ) then
            local skey = ( type( key ) == "string" ) and string.format( "[ %q ]", key ) or string.format( "[ %d ]", key )
            if type( tbl[ key ] ) == "table" then
                sortserialize( tbl[ key ], skey, file, tab .. "    " )
                file:write( ",\n" )
            else
                local svalue = ( type( tbl[ key ] ) == "string" ) and string.format( "%q", tbl[ key ] ) or tostring( tbl[ key ] )
                file:write( tab .. "    " .. skey .. " = " .. svalue )
                file:write( ",\n" )
            end
        end
    end
    file:write( "\n" )
    file:write( tab .. "}" )
end

savetable = function( tbl, name, path )
    local file, err = io.open( path, "w+" )
    if file then
        file:write( "local " .. name .. "\n\n" )
        sortserialize( tbl, name, file, "" )
        file:write( "\n\nreturn " .. name )
        file:close( )
        return true
    else
        return false, err
    end
end

loadtable = function( path )
    local file, err = io.open( path, "r" )
    if not file then
        return nil, err
    end
    local content = file:read "*a"
    file:close( )
    local chunk, err = loadstring( content )
    if chunk then
        local ret = chunk( )
        if ret and type( ret ) == "table" then
            return ret
        else
            return nil, "invalid table"
        end
    end
    return nil, err
end

savearray = function( array, path )
    array = array or { }
    local file, err = io.open( path, "w+" )
    if not file then
        return false, err
    end
    local iterate, savetbl
    iterate = function( tbl )
        local tmp = { }
        for key, value in pairs( tbl ) do
            tmp[ #tmp + 1 ] = tostring( key )
        end
        table.sort( tmp )
        for i, key in ipairs( tmp ) do
            key = tonumber( key ) or key
            if type( tbl[ key ] ) == "table" then
                file:write( ( ( type( key ) ~= "number" ) and tostring( key ) .. " = " ) or " " )
                savetbl( tbl[ key ] )
            else
                file:write( ( ( type( key ) ~= "number" and tostring( key ) .. " = " ) or "" ) .. ( ( type( tbl[ key ] ) == "string" ) and string.format( "%q", tbl[ key ] ) or tostring( tbl[ key ] ) ) .. ", " )
            end
        end
    end
    savetbl = function( tbl )
        local tmp = { }
        for key, value in pairs( tbl ) do
            tmp[ #tmp + 1 ] = tostring( key )
        end
        table.sort( tmp )
        file:write( "{ " )
        iterate( tbl )
        file:write( "}, " )
    end
    file:write( "return {\n\n" )
    for i, tbl in ipairs( array ) do
        if type( tbl ) == "table" then
            file:write( "    { " )
            iterate( tbl )
            file:write( "},\n" )
        else
            file:write( "    " .. string.format( "%q", tostring( tbl ) ) .. ",\n" )
        end
    end
    file:write( "\n}" )
    file:close( )
    return true
end

formatseconds = function( t )
    return
        math_floor( t / ( 60 * 60 * 24 ) ),
        math_floor( t / ( 60 * 60 ) ) % 24,
        math_floor( t / 60 ) % 60,
        t % 60
end

--// convert bytes to the right unit  / based on a function by Night
formatbytes = function( bytes )
    local bytes = tonumber( bytes )
    if ( not bytes ) or ( not type( bytes ) == "number" ) or ( bytes < 0 ) or ( bytes == 1 / 0 ) then
        --return nil, "util.lua: error in formatbytes(), parameter not valid"
        return nil
    end
    if bytes == 0 then return "0 B" end
    local i, units = 1, { "B", "KB", "MB", "GB", "TB", "PB", "EB", "YB" }
    while bytes >= 1024 do
        bytes = bytes / 1024
        i = i + 1
    end
    if units[ i ] == "B" then
        return string_format( "%.0f", bytes ) .. " " .. ( units[ i ] or "?" )
    else
        return string_format( "%.2f", bytes ) .. " " .. ( units[ i ] or "?" )
    end
end

generatepass = function( len )
    local len = tonumber( len )
    if not ( type( len ) == "number" ) or ( len < 0 ) or ( len > 1000 ) then len = 20 end
    local lower = { "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m",
                    "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z" }
    local upper = { "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
                    "N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z" }
    math_randomseed( os_time() )
    local pwd = ""
    for i = 1, len do
        local X = math_random( 0, 9 )
        if X < 4 then
            pwd = pwd .. math_random( 0, 9 )
        elseif ( X >= 4 ) and ( X < 6 ) then
            pwd = pwd .. upper[ math_random( 1, 25 ) ]
        else
            pwd = pwd .. lower[ math_random( 1, 25 ) ]
        end
    end
    return pwd
end

--// returns current date in new luadch date style: yyyymmddhhmmss (as number)
date = function()
    local year = os_date( "%Y" )
    local month = os_date( "%m" )
    local day = os_date( "%d" )
    local hour = os_date( "%H" )
    local minutes = os_date( "%M" )
    local seconds = os_date( "%S" )
    return tonumber( year .. month .. day .. hour .. minutes .. seconds )
end

--// returns difftime between two date values (new luadch date style)
difftime = function( t1, t2 )
    local err
    if not t1 then
        err = "util.lua: error in param #1: got nil"
        return nil, err
    end
    if not t2 then
        err = "util.lua: error in param #2: got nil"
        return nil, err
    end
    if type( t1 ) ~= "number" then
        err = "util.lua: error in param #1: number expected, got " .. type( t1 )
        return nil, err
    end
    if type( t2 ) ~= "number" then
        err = "util.lua: error in param #2: number expected, got " .. type( t2 )
        return nil, err
    end
    local t1, t2 = tostring( t1 ), tostring( t2 )
    local y1, m1, d1, h1, M1, s1
    local y2, m2, d2, h2, M2, s2
    local diff, T1, T2
    local y, d, h, m, s
    if #t1 ~= 14 then
        err = "util.lua: error in param #1: not valid"
        return nil, err
    else
        y1 = t1:sub( 1, 4 )
        m1 = t1:sub( 5, 6 )
        d1 = t1:sub( 7, 8 )
        h1 = t1:sub( 9, 10 )
        M1 = t1:sub( 11, 12 )
        s1 = t1:sub( 13, 14 )
    end
    if #t2 ~= 14 then
        err = "util.lua: error in param #2: not valid"
        return nil, err
    else
        y2 = t2:sub( 1, 4 )
        m2 = t2:sub( 5, 6 )
        d2 = t2:sub( 7, 8 )
        h2 = t2:sub( 9, 10 )
        M2 = t2:sub( 11, 12 )
        s2 = t2:sub( 13, 14 )
    end
    T1 = os_time( { year = y1, month = m1, day = d1, hour = h1, min = M1, sec = s1 } )
    T2 = os_time( { year = y2, month = m2, day = d2, hour = h2, min = M2, sec = s2 } )
    diff = os_difftime( T1, T2 )
    y = math_floor( diff / ( 60 * 60 * 24 ) / 365 )
    d = math_floor( diff / ( 60 * 60 * 24 ) ) % 365
    h = math_floor( diff / ( 60 * 60 ) ) % 24
    m = math_floor( diff / 60 ) % 60
    s = diff % 60
    return y, d, h, m, s
end

--// convert os.time() "epoch" date to luadch date style: yyyymmddhhmmss (as number)
convertepochdate = function( t )
    local err
    if type( t ) ~= "number" then
        err = "util.lua: error: number expected, got " .. type( t )
        return nil, err
    end
    local date, y, m, d, h, M, s
    date = os_date( "*t", t )
    y = tostring( date.year )
    m = tostring( date.month )
    d = tostring( date.day )
    h = tostring( date.hour )
    M = tostring( date.min )
    s = tostring( date.sec )
    if #m == 1 then m = "0" .. m end
    if #d == 1 then d = "0" .. d end
    if #h == 1 then h = "0" .. h end
    if #M == 1 then M = "0" .. M end
    if #s == 1 then s = "0" .. s end
    return tonumber( y .. m .. d .. h .. M .. s ), err
end

--// trim whitespaces from both ends of a string
trimstring = function( str )
    local err
    local str = tostring( str )
    if type( str ) ~= "string" then
        err = "util.lua: error: string expected, got " .. type( str )
        return nil, err
    end
    return string_find( str, "^%s*$" ) and "" or string_match( str, "^%s*(.*%S)" )
end

string2table = function( str )
    local chunk, err = loadstring( str )
    if chunk then
        local ret = chunk( )
        if ret and type( ret ) == "table" then
            return ret
        else
            return nil, "invalid table"
        end
    end
    return nil, err
end

return {

    savetable = savetable,
    loadtable = loadtable,
    savearray = savearray,
    formatseconds = formatseconds,
    formatbytes = formatbytes,
    generatepass = generatepass,
    date = date,
    difftime = difftime,
    convertepochdate = convertepochdate,
    trimstring = trimstring,
    string2table = string2table,

}

system/core/twodns.lua
--[[

    twodns.lua by pulsar

        usage:

            table/nil, statuscode = twodns.domains( https )
            table/nil, statuscode = twodns.accinfo( https, email, token )

]]


--// caching table lookups
local string_gmatch = string.gmatch

--// functions
local domains, accinfo

--// return available domains as array
domains = function( https )
    local address = "https://api.twodns.de/domains"
    local body, code, headers, status = https.request( address )
    if tostring( code ) == "200" then
        body = body:gsub('%[',''):gsub('%]',''):gsub('{',''):gsub('}',''):gsub('"',''):gsub('name:',''):gsub(',',' ')
        t, i = {}, 1
        for v in string_gmatch( body, "(%S+)" ) do
            t[ i ] = v
            i = i + 1
        end
        return t, code
    end
    return nil, code
end

--// return users complete account info as table
accinfo = function( https, email, token )
    local string2table = function( str )
        local chunk, err = loadstring( str )
        if chunk then
            local ret = chunk()
            if ret and type( ret ) == "table" then
                return ret
            else
                return nil, "invalid table"
            end
        end
        return nil, err
    end
    local response = {}
    local save = ltn12.sink.table( response )
    local body, code, headers, status = https.request{
        url = "https://api.twodns.de/users/me",
        method = "GET",
        sink = save,
        headers = { authorization = "Basic " .. ( mime.b64( email .. ":" .. token ) ), }
    }
    if tostring( code ) == "200" then
        local sink_data = response[ 1 ]
        local sink_len = sink_data:len()
        local sink_data = sink_data:sub( 2, sink_len - 1 )
        local sink_str = sink_data:gsub('%[','{'):gsub('%]','}'):gsub(':','='):gsub('"=','" ] = '):gsub('{"','{ %[ "'):gsub(',"',',%[ "'):gsub('"api_token','%[ "api_token')
        local t_header = "local twodns\n\ntwodns = {\n\n"
        local t_footer = "\n\n}\n\nreturn twodns"
        local t, err = string2table( t_header .. sink_str .. t_footer )
        if t then
            return t, code
        else
            return nil, err
        end
    end
    return nil, code
end

return {

    domains = domains,
    accinfo = accinfo,

}

system/modules/lualibs/socket/http.lua
-----------------------------------------------------------------------------
-- HTTP/1.1 client support for the Lua language.
-- LuaSocket toolkit.
-- Author: Diego Nehab
-- RCS ID: $Id: http.lua,v 1.71 2007/10/13 23:55:20 diego Exp $
-----------------------------------------------------------------------------

-----------------------------------------------------------------------------
-- Declare module and import dependencies
-------------------------------------------------------------------------------
local socket = require("socket")
local url = require("socket.url")
local ltn12 = require("ltn12")
local mime = require("mime")
local string = require("string")
local base = _G
local table = require("table")
module("socket.http")

-----------------------------------------------------------------------------
-- Program constants
-----------------------------------------------------------------------------
-- connection timeout in seconds
TIMEOUT = 60
-- default port for document retrieval
PORT = 80
-- user agent field sent in request
USERAGENT = "Mozilla/5.0 (Windows NT 6.1; WOW64; Trident/7.0; AS; rv:11.0) like Gecko"

-----------------------------------------------------------------------------
-- Reads MIME headers from a connection, unfolding where needed
-----------------------------------------------------------------------------
local function receiveheaders(sock, headers)
    local line, name, value, err
    headers = headers or {}
    -- get first line
    line, err = sock:receive()
    if err then return nil, err end
    -- headers go until a blank line is found
    while line ~= "" do
        -- get field-name and value
        name, value = socket.skip(2, string.find(line, "^(.-):%s*(.*)"))
        if not (name and value) then return nil, "malformed reponse headers" end
        name = string.lower(name)
        -- get next line (value might be folded)
        line, err  = sock:receive()
        if err then return nil, err end
        -- unfold any folded values
        while string.find(line, "^%s") do
            value = value .. line
            line = sock:receive()
            if err then return nil, err end
        end
        -- save pair in table
        if headers[name] then headers[name] = headers[name] .. ", " .. value
        else headers[name] = value end
    end
    return headers
end

-----------------------------------------------------------------------------
-- Extra sources and sinks
-----------------------------------------------------------------------------
socket.sourcet["http-chunked"] = function(sock, headers)
    return base.setmetatable({
        getfd = function() return sock:getfd() end,
        dirty = function() return sock:dirty() end
    }, {
        __call = function()
            -- get chunk size, skip extention
            local line, err = sock:receive()
            if err then return nil, err end
            local size = base.tonumber(string.gsub(line, ";.*", ""), 16)
            if not size then return nil, "invalid chunk size" end
            -- was it the last chunk?
            if size > 0 then
                -- if not, get chunk and skip terminating CRLF
                local chunk, err, part = sock:receive(size)
                if chunk then sock:receive() end
                return chunk, err
            else
                -- if it was, read trailers into headers table
                headers, err = receiveheaders(sock, headers)
                if not headers then return nil, err end
            end
        end
    })
end

socket.sinkt["http-chunked"] = function(sock)
    return base.setmetatable({
        getfd = function() return sock:getfd() end,
        dirty = function() return sock:dirty() end
    }, {
        __call = function(self, chunk, err)
            if not chunk then return sock:send("0\r\n\r\n") end
            local size = string.format("%X\r\n", string.len(chunk))
            return sock:send(size ..  chunk .. "\r\n")
        end
    })
end

-----------------------------------------------------------------------------
-- Low level HTTP API
-----------------------------------------------------------------------------
local metat = { __index = {} }

function open(host, port, create)
    -- create socket with user connect function, or with default
    local c = socket.try((create or socket.tcp)())
    local h = base.setmetatable({ c = c }, metat)
    -- create finalized try
    h.try = socket.newtry(function() h:close() end)
    -- set timeout before connecting
    h.try(c:settimeout(TIMEOUT))
    h.try(c:connect(host, port or PORT))
    -- here everything worked
    return h
end

function metat.__index:sendrequestline(method, uri)
    local reqline = string.format("%s %s HTTP/1.1\r\n", method or "GET", uri)
    return self.try(self.c:send(reqline))
end

function metat.__index:sendheaders(headers)
    local h = "\r\n"
    for i, v in base.pairs(headers) do
        h = i .. ": " .. v .. "\r\n" .. h
    end
    self.try(self.c:send(h))
    return 1
end

function metat.__index:sendbody(headers, source, step)
    source = source or ltn12.source.empty()
    step = step or ltn12.pump.step
    -- if we don't know the size in advance, send chunked and hope for the best
    local mode = "http-chunked"
    if headers["content-length"] then mode = "keep-open" end
    return self.try(ltn12.pump.all(source, socket.sink(mode, self.c), step))
end

function metat.__index:receivestatusline()
    local status = self.try(self.c:receive(5))
    -- identify HTTP/0.9 responses, which do not contain a status line
    -- this is just a heuristic, but is what the RFC recommends
    if status ~= "HTTP/" then return nil, status end
    -- otherwise proceed reading a status line
    status = self.try(self.c:receive("*l", status))
    local code = socket.skip(2, string.find(status, "HTTP/%d*%.%d* (%d%d%d)"))
    return self.try(base.tonumber(code), status)
end

function metat.__index:receiveheaders()
    return self.try(receiveheaders(self.c))
end

function metat.__index:receivebody(headers, sink, step)
    sink = sink or ltn12.sink.null()
    step = step or ltn12.pump.step
    local length = base.tonumber(headers["content-length"])
    local t = headers["transfer-encoding"] -- shortcut
    local mode = "default" -- connection close
    if t and t ~= "identity" then mode = "http-chunked"
    elseif base.tonumber(headers["content-length"]) then mode = "by-length" end
    return self.try(ltn12.pump.all(socket.source(mode, self.c, length),
        sink, step))
end

function metat.__index:receive09body(status, sink, step)
    local source = ltn12.source.rewind(socket.source("until-closed", self.c))
    source(status)
    return self.try(ltn12.pump.all(source, sink, step))
end

function metat.__index:close()
    return self.c:close()
end

-----------------------------------------------------------------------------
-- High level HTTP API
-----------------------------------------------------------------------------
local function adjusturi(reqt)
    local u = reqt
    -- if there is a proxy, we need the full url. otherwise, just a part.
    if not reqt.proxy and not PROXY then
        u = {
           path = socket.try(reqt.path, "invalid path 'nil'"),
           params = reqt.params,
           query = reqt.query,
           fragment = reqt.fragment
        }
    end
    return url.build(u)
end

local function adjustproxy(reqt)
    local proxy = reqt.proxy or PROXY
    if proxy then
        proxy = url.parse(proxy)
        return proxy.host, proxy.port or 3128
    else
        return reqt.host, reqt.port
    end
end

local function adjustheaders(reqt)
    -- default headers
    local lower = {
        ["user-agent"] = USERAGENT,
        ["host"] = reqt.host,
        ["connection"] = "close, TE",
        ["te"] = "trailers"
    }
    -- if we have authentication information, pass it along
    if reqt.user and reqt.password then
        lower["authorization"] = 
            "Basic " ..  (mime.b64(reqt.user .. ":" .. reqt.password))
    end
    -- override with user headers
    for i,v in base.pairs(reqt.headers or lower) do
        lower[string.lower(i)] = v
    end
    return lower
end

-- default url parts
local default = {
    host = "",
    port = PORT,
    path ="/",
    scheme = "http"
}

local function adjustrequest(reqt)
    -- parse url if provided
    local nreqt = reqt.url and url.parse(reqt.url, default) or {}
    -- explicit components override url
    for i,v in base.pairs(reqt) do nreqt[i] = v end
    if nreqt.port == "" then nreqt.port = 80 end
    socket.try(nreqt.host and nreqt.host ~= "", 
        "invalid host '" .. base.tostring(nreqt.host) .. "'")
    -- compute uri if user hasn't overriden
    nreqt.uri = reqt.uri or adjusturi(nreqt)
    -- ajust host and port if there is a proxy
    nreqt.host, nreqt.port = adjustproxy(nreqt)
    -- adjust headers in request
    nreqt.headers = adjustheaders(nreqt)
    return nreqt
end

local function shouldredirect(reqt, code, headers)
    return headers.location and
           string.gsub(headers.location, "%s", "") ~= "" and
           (reqt.redirect ~= false) and
           (code == 301 or code == 302) and
           (not reqt.method or reqt.method == "GET" or reqt.method == "HEAD")
           and (not reqt.nredirects or reqt.nredirects < 5)
end

local function shouldreceivebody(reqt, code)
    if reqt.method == "HEAD" then return nil end
    if code == 204 or code == 304 then return nil end
    if code >= 100 and code < 200 then return nil end
    return 1
end

-- forward declarations
local trequest, tredirect

function tredirect(reqt, location)
    local result, code, headers, status = trequest {
        -- the RFC says the redirect URL has to be absolute, but some
        -- servers do not respect that
        url = url.absolute(reqt.url, location),
        source = reqt.source,
        sink = reqt.sink,
        headers = reqt.headers,
        proxy = reqt.proxy, 
        nredirects = (reqt.nredirects or 0) + 1,
        create = reqt.create
    }   
    -- pass location header back as a hint we redirected
    headers = headers or {}
    headers.location = headers.location or location
    return result, code, headers, status
end

function trequest(reqt)
    -- we loop until we get what we want, or
    -- until we are sure there is no way to get it
    local nreqt = adjustrequest(reqt)
    local h = open(nreqt.host, nreqt.port, nreqt.create)
    -- send request line and headers
    h:sendrequestline(nreqt.method, nreqt.uri)
    h:sendheaders(nreqt.headers)
    -- if there is a body, send it
    if nreqt.source then
        h:sendbody(nreqt.headers, nreqt.source, nreqt.step) 
    end
    local code, status = h:receivestatusline()
    -- if it is an HTTP/0.9 server, simply get the body and we are done
    if not code then
        h:receive09body(status, nreqt.sink, nreqt.step)
        return 1, 200
    end
    local headers
    -- ignore any 100-continue messages
    while code == 100 do 
        headers = h:receiveheaders()
        code, status = h:receivestatusline()
    end
    headers = h:receiveheaders()
    -- at this point we should have a honest reply from the server
    -- we can't redirect if we already used the source, so we report the error 
    if shouldredirect(nreqt, code, headers) and not nreqt.source then
        h:close()
        return tredirect(reqt, headers.location)
    end
    -- here we are finally done
    if shouldreceivebody(nreqt, code) then
        h:receivebody(headers, nreqt.sink, nreqt.step)
    end
    h:close()
    return 1, code, headers, status
end

local function srequest(u, b)
    local t = {}
    local reqt = {
        url = u,
        sink = ltn12.sink.table(t)
    }
    if b then
        reqt.source = ltn12.source.string(b)
        reqt.headers = {
            ["content-length"] = string.len(b),
            ["content-type"] = "application/x-www-form-urlencoded"
        }
        reqt.method = "POST"
    end
    local code, headers, status = socket.skip(1, trequest(reqt))
    return table.concat(t), code, headers, status
end

request = socket.protect(function(reqt, body)
    if base.type(reqt) == "string" then return srequest(reqt, body)
    else return trequest(reqt) end
end)
